tests/components/broadlink/test_device.py

```diff
@@ -0,0 +1,389 @@
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ) as mock_forward, patch.object(
+        hass.config_entries.flow, "async_init"
+    ) as mock_init:
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_SETUP_RETRY
+    assert mock_api.auth.call_count == 1
+    assert mock_forward.call_count == 0
+    assert mock_init.call_count == 0
+
+
+async def test_device_setup_broadlink_exception(hass):
+    """Test we handle a Broadlink exception."""
+    device = get_device("Office")
+    mock_api = device.get_mock_api()
+    mock_api.auth.side_effect = blke.BroadlinkException()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ) as mock_forward, patch.object(
+        hass.config_entries.flow, "async_init"
+    ) as mock_init:
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_SETUP_ERROR
+    assert mock_api.auth.call_count == 1
+    assert mock_forward.call_count == 0
+    assert mock_init.call_count == 0
+
+
+async def test_device_setup_update_device_offline(hass):
+    """Test we handle a device offline in the update step."""
+    device = get_device("Office")
+    mock_api = device.get_mock_api()
+    mock_api.check_sensors.side_effect = blke.DeviceOfflineError()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ) as mock_forward, patch.object(
+        hass.config_entries.flow, "async_init"
+    ) as mock_init:
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_SETUP_RETRY
+    assert mock_api.auth.call_count == 1
+    assert mock_api.check_sensors.call_count == 1
+    assert mock_forward.call_count == 0
+    assert mock_init.call_count == 0
+
+
+async def test_device_setup_update_authorization_error(hass):
+    """Test we handle an authorization error in the update step."""
+    device = get_device("Office")
+    mock_api = device.get_mock_api()
+    mock_api.check_sensors.side_effect = (blke.AuthorizationError(), None)
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ) as mock_forward, patch.object(
+        hass.config_entries.flow, "async_init"
+    ) as mock_init:
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_LOADED
+    assert mock_api.auth.call_count == 2
+    assert mock_api.check_sensors.call_count == 2
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6042836)